File: packages/silo-client/src/cache/CacheKey.ts
import { QueryString } from "../transport/query-string.js";
import type { TransportQueryValue } from "../transport/transport-request.js";

/** Request identity within a transport whose URL and authentication are fixed. */
export class CacheKey {
  static of(method: string, path: string, query?: Record<string, TransportQueryValue>): string {
    const sorted = query && Object.fromEntries(Object.keys(query).sort().map((name) => [name, query[name]]));
    return `${method} ${path}${QueryString.build(sorted)}`;
  }

  static matches(key: string, pathPrefix: string): boolean {
    const path = key.slice(key.indexOf(" ") + 1);
    return path === pathPrefix || path.startsWith(`${pathPrefix}/`) || path.startsWith(`${pathPrefix}?`);
  }
}

File: packages/silo-client-dart/lib/src/cache/cache_key.dart
import '../transport/query_string.dart';

/// A response's key: the method, the path and every query parameter, sorted.
abstract final class CacheKey {
  static String of(String method, String path, Map<String, Object?> query) {
    final sorted = {for (final name in query.keys.toList()..sort()) name: query[name]};
    return '$method $path${QueryString.build(sorted)}';
  }

  /// Whether a write to [pathPrefix] makes [key] stale, without matching a
  /// sibling such as `posts-archive`.
  static bool matches(String key, String pathPrefix) {
    final target = key.substring(key.indexOf(' ') + 1);
    return target == pathPrefix || target.startsWith('$pathPrefix/') || target.startsWith('$pathPrefix?');
  }
}
